services/wireguard/service.go

```diff
@@ -27,35 +28,46 @@ import (
 	"github.com/mysteriumnetwork/node/identity"
 	"github.com/mysteriumnetwork/node/money"
 	dto_discovery "github.com/mysteriumnetwork/node/service_discovery/dto"
+	"github.com/mysteriumnetwork/node/services/wireguard/resources"
 	"github.com/mysteriumnetwork/node/session"
 )
 
 const logPrefix = "[service-wireguard] "
 
-// ErrAlreadyStarted is the error we return when the start is called multiple times
+// ErrAlreadyStarted is the error we return when the start is called multiple times.
 var ErrAlreadyStarted = errors.New("Service already started")
 
 // NewManager creates new instance of Wireguard service
 func NewManager(locationResolver location.Resolver, ipResolver ip.Resolver) *Manager {
 	return &Manager{
 		locationResolver: locationResolver,
 		ipResolver:       ipResolver,
+		resources:        resources.Handler{},
 	}
 }
 
-// Manager represents entrypoint for Wireguard service
+// Manager represents entrypoint for Wireguard service.
 type Manager struct {
-	process          sync.WaitGroup
-	locationResolver location.Resolver
-	ipResolver       ip.Resolver
-	isStarted        bool
+	isStarted          bool
+	process            sync.WaitGroup
+	locationResolver   location.Resolver
+	ipResolver         ip.Resolver
+	connectionEndpoint ConfigProvider
+	resources          resources.Handler
 }
 
-// Config represent a Wireguard service provider configuration that will be passed to the consumer for establishing a connection
+// Config represent a Wireguard service provider configuration that will be passed to the consumer for establishing a connection.
 type Config struct {
-	PublicKey string
-	IP        string
-	Endpoint  string
+	Provider Provider
+	Consumer Consumer
+}
+
+// ConfigProvider represents Wireguard network instance, it provide information
+// required for establishing connection between service provider and consumer.
+type ConfigProvider interface {
+	ProviderConfig(ip net.IPNet) (Provider, error)
```

Comment:
ConfigProvider method ProviderConfig returns Provider which is actually configuration of service provider. As per previous suggestion why is bad to call Provider struct what it actually is ? ProviderConfiguration, the same with Consumer -> ConsumerConfiguration. Another question - why ip parameter is needed to provide config - maybe it either can be set in actual interface implementor or other parameters should be exposed too. Actually I don't understand purpose of this interface, especially Close method (Config provider can be closed, does that imply that some kind of resources need to be released?)

And actual config provider implementation can take resources as depedency and do all that allocation inside. That way details wont be exposed in service manager itself (i.e. IP passing)

Reply:
The problem here is that `ProviderConfig` is not just config generation, it also starts service provider to listen for connections. Same for `ConsumerConfig`, it adds the consumer to the allowed list and returns configuration for the consumer. 

The IP address is required to do this configuration. 

`Close` release interfaces and stops service. 

I'll change it to avoid doing two things in one place.

Reply:
So maybe thats the problem - component does too much unrelated work, which can be split? If we have connection - let's handle connection details there, if we need some details for connection and those details need also to be shared with consumer - let's make configuration item and pass it both to consumer and to that wg connnection

Reply:
Changed it to the things with a single responsibility.